llvm/lib/CASObjectFormats/SimpleV1.cpp

```diff
@@ -0,0 +1,1217 @@
+//===- CASObjectFormats/SimpleV1.cpp --------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/CASObjectFormats/SimpleV1.h"
+#include "llvm/ADT/PointerUnion.h"
+#include "llvm/ADT/PostOrderIterator.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/CAS/CASDB.h"
+#include "llvm/CASObjectFormats/Data.h"
+#include "llvm/CASObjectFormats/Encoding.h"
+#include "llvm/CASObjectFormats/ObjectFormatHelpers.h"
+#include "llvm/ExecutionEngine/JITLink/JITLink.h"
+#include "llvm/Support/EndianStream.h"
+#include <memory>
+
+// FIXME: For cl::opt. Should thread through or delete.
+#include "llvm/Support/CommandLine.h"
+
+using namespace llvm;
+using namespace llvm::casobjectformats;
+using namespace llvm::casobjectformats::simplev1;
+
+const StringLiteral CompileUnitRef::KindString;
+const StringLiteral NameRef::KindString;
+const StringLiteral SectionRef::KindString;
+const StringLiteral BlockRef::KindString;
+const StringLiteral SymbolRef::KindString;
+const StringLiteral EdgeListRef::KindString;
+
+void ObjectFileSchema::anchor() {}
+
+Expected<cas::NodeRef>
+ObjectFileSchema::createFromLinkGraphImpl(const jitlink::LinkGraph &G,
+                                          raw_ostream *DebugOS) const {
+  return CompileUnitRef::create(*this, G, DebugOS);
+}
+
+Expected<std::unique_ptr<jitlink::LinkGraph>>
+ObjectFileSchema::createLinkGraphImpl(
+    cas::NodeRef RootNode, StringRef Name,
+    jitlink::LinkGraph::GetEdgeKindNameFunction GetEdgeKindName,
+    raw_ostream *) const {
+  if (!isRootNode(RootNode))
+    return createStringError(inconvertibleErrorCode(), "invalid root node");
+  auto CU = CompileUnitRef::get(*this, RootNode);
+  if (!CU)
+    return CU.takeError();
+  return CU->createLinkGraph(Name, GetEdgeKindName);
+}
+
+ObjectFileSchema::ObjectFileSchema(cas::CASDB &CAS) : SchemaBase(CAS) {
+  // Fill the cache immediately to preserve thread-safety.
+  if (Error E = fillCache())
+    report_fatal_error(std::move(E));
+}
+
+Error ObjectFileSchema::fillCache() {
+  Optional<cas::CASID> RootKindID;
+  const unsigned Version = 0; // Bump this to error on old object files.
+  if (Expected<cas::NodeRef> ExpectedRootKind =
+          CAS.createNode(None, "cas.o:simplev1:schema:" + Twine(Version).str()))
+    RootKindID = *ExpectedRootKind;
+  else
+    return ExpectedRootKind.takeError();
+
+  StringRef AllKindStrings[] = {
+      BlockRef::KindString,  CompileUnitRef::KindString,
+      NameRef::KindString,   SectionRef::KindString,
+      SymbolRef::KindString, EdgeListRef::KindString,
+  };
```

Comment:
I was expecting only two things here:
- CompileUnitRef
- ContentRef (or BlockContentRef or something)

Beyond that:
- I could be convinced to include fixups in the (Block)ContentRef (but not embedded)
- I could be convinced to have a NameRef (since names are big and probably repeat)

Otherwise I think the rest should go!